nuxeo-core/nuxeo-core-event/src/main/java/org/nuxeo/ecm/core/event/pipe/queue/QueueBaseEventBundlePipe.java

```diff
@@ -0,0 +1,148 @@
+            try {
+                batchSize = Integer.parseInt(params.get(batchSize));
+            } catch (NumberFormatException e) {
+                log.error("Unable to read batchSize parameter", e);
+            }
+        }
+        queue = new ConcurrentLinkedQueue<>();
+        consumerTPE = new ThreadPoolExecutor(1, 1, 60, TimeUnit.MINUTES, new LinkedBlockingQueue<>());
+        consumerTPE.prestartCoreThread();
+        consumerTPE.execute(new Runnable() {
+
+            protected boolean send(List<EventBundle> messages) {
+                if (consumer.receiveMessage(messages)) {
+                    messages.clear();
+                } else {
+                    // keep the events that can not be processed ?
+                    queue.addAll(messages);
+                    return false;
+                }
+                return true;
+            }
+
+            @Override
+            public void run() {
+
+                consumer = new LocalEventBundlePipeConsumer();
+                consumer.initConsumer(getName(), getParameters());
+                while (!stop) {
+                    List<EventBundle> messages = new ArrayList<>();
+                    EventBundle message;
+                    while ((message = queue.poll()) != null) {
+                        messages.add(message);
+                        if (messages.size() >= batchSize) {
+                            send(messages);
+                        }
+                    }
+                    if (messages.size() > 0) {
+                        send(messages);
+                    }
+
+                    // XXX this is a hack ! TODO: find a better approach
+                    try {
+                        if (Framework.isTestModeSet()) {
+                            Thread.sleep(5);
+                        } else {
+                            Thread.sleep(200);
+                        }
+                    } catch (InterruptedException e) {
+                        consumerTPE.shutdown();
+                    }
+                }
+
+            }
+        });
+    }
+
+    @Override
+    protected EventBundle marshall(EventBundle events) {
+        return events;
+    }
+
+    @Override
+    protected void send(EventBundle message) {
+        queue.offer(message);
+    }
+
+    @Override
+    public void shutdown() throws InterruptedException {
+        stop = true;
+        // XXX
+        waitForCompletion(5000L);
+        if (consumer != null) {
+            consumer.shutdown();
+        }
+    }
+
+    @Override
+    public boolean waitForCompletion(long timeoutMillis) throws InterruptedException {
+        long deadline = System.currentTimeMillis() + timeoutMillis;
+        int pause = (int) Math.min(timeoutMillis, 500L);
+        do {
+            if (queue.size() == 0) {
+                return true;
+            }
+            Thread.sleep(pause);
+        } while (System.currentTimeMillis() < deadline);
```

Comment:
It would be better to use a Lock + Condition to notify of empty queue. See RedisBlockQueue for example.